api/TransportTypes.ts

```diff
@@ -0,0 +1,131 @@
+import { JSONSerializableValue } from "lib/Types/HelperTypes";
+import { MatchFnCollection } from "lib/Types/MatchType";
+import { StrictExtends } from "lib/Types/StrictExtendsType";
+import { Endpoint, URLParameters } from "lib/URL";
+import { ZodType, ZodTypeAny, z } from "zod";
+import { Middleware, chainMiddleware } from "./Middleware";
+import { tryParseAPICall } from "./Validation";
+
+export type TiFHTTPMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE"
+
+export type TiFAPIRequestBody = { [key: string]: JSONSerializableValue }
+
+export type TiFAPINoContentSchema = "no-content"
+
+type StatusCodeMap = {
+  status200: 200
+  status201: 201
+  status204: 204
+  status400: 400
+  status401: 401
+  status403: 403
+  status404: 404
+  status429: 429
+  status500: 500
+}
+
+/**
+ * A type mapping a zod schema to an http status code. The keys of the object
+ * must be in the form `statusXXX`.
+ */
+export type TiFAPIResponseSchemas = Partial<{
+  [key in keyof StatusCodeMap]: key extends "status204"
+    ? TiFAPINoContentSchema
+    : ZodType
+}>
+
+const EmptyObjectSchema = z.object({})
+
+type SchemaFor<
+  Key extends keyof StatusCodeMap,
+  Schemas extends TiFAPIResponseSchemas
+> = Key extends "status204"
+  ? Schemas[Key] extends TiFAPINoContentSchema
+    ? typeof EmptyObjectSchema
+    : undefined
+  : Schemas[Key]
+
+type TiFAPIResponse<Schemas extends TiFAPIResponseSchemas> = {
+  [key in keyof StatusCodeMap]: SchemaFor<key, Schemas> extends ZodType
+    ? {
+        status: StatusCodeMap[key]
+        data: z.infer<SchemaFor<key, Schemas>>
+      }
+    : never
+}[keyof StatusCodeMap]
+
+export type GenericEndpointSchema = EndpointSchema<InputSchema | Omit<InputSchema, "body"> | {}, any>
+
+type TiFAPIInput = {
+  body?: TiFAPIRequestBody;
+  query?: URLParameters;
+  params?: URLParameters;
+  signal?: AbortSignal
+} | void;
+
+export type GenericEndpointFunction = (_: TiFAPIInput) => Promise<TiFAPIResponse<any>>
+
+/**
+ * Transforms endpoint schemas into middleware for API implementations.
+ */
+export type EndpointSchemaToMiddleware = (endpointName: string, endpointSchema: GenericEndpointSchema) => Middleware<TiFAPIInput, TiFAPIResponse<any>>;
+
+/**
+ * Type asserts an endpoint schema.
+ */
+export const assertEndpointSchemaType = <TInput extends InputSchema, TOutput extends TiFAPIResponseSchemas>(
+  endpoint: EndpointSchema<
+    StrictExtends<InputSchema, TInput>, 
+    StrictExtends<TiFAPIResponseSchemas, TOutput>
+  >
+) => endpoint 
+
+/**
+ * Map of endpoint schemas.
+ */
+export type APISchema = Record<string, EndpointSchema<any, any>>
+
+/**
+ * Transforms a map of endpoint schemas to a map of functions where the inputs and outputs are inferred from the endpoint schemas.
+ */
+export type EndpointSchemasToFunctions<T extends APISchema> = { [K in keyof T]: (_: {} extends T[K]['input'] ? void : TiFAPIRequest<T[K]['input']> & {signal?: AbortSignal}) => Promise<TiFAPIResponse<T[K]['outputs']>>}
```

Comment:
I love how this became a masterclass on typescript types... The truly beautiful core that I started this with has now expanded into something akin to the Mona Lisa. What an incredible work of art, and we should contact a few museums to display it starting with the Vatican. This is a masterpiece that all future generations must know of.